### PR TITLE
Add JIT support

### PR DESCRIPTION
**Pull request by roastduck**
Here JIT means automatic recompilation if stage-1 parameter has changed. See the updated doc for details.

Currently JIT is not supported for AD, because JIT is triggered by users passing arguments, and users need the parameter map returned from AD to pass arguments.

### CODE REVIEW

**Review by Blealtan — APPROVED** [review]

File: python/freetensor/core/jit.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+from typing import Tuple, Callable, Mapping, Any
+
+
+class JITMeta(type):
+
+    def __init__(self, *args, **kvs):
+        super(JITMeta, self).__init__(*args, **kvs)
+        self.inner_type = None
+
+    def __getitem__(self, inner_type):
```
[the remark this answers is not in the document]
Comment (roastduck) [reply python/freetensor/core/jit.py]: Fixed

File: python/freetensor/core/jit.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
+import abc
+import copy
+import inspect
```
[the remark this answers is not in the document]
Comment (roastduck) [reply python/freetensor/core/jit.py]: Fixed

File: python/freetensor/core/jit.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+from typing import Tuple, Callable, Mapping, Any
```
[the remark this answers is not in the document]
Comment (roastduck) [reply python/freetensor/core/jit.py]: Fixed

File: python/freetensor/core/jit.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+
+class JITMeta(type):
+
+    def __init__(self, *args, **kvs):
```
[the remark this answers is not in the document]
Comment (roastduck) [reply python/freetensor/core/jit.py]: Fixed

**Pull request merged**